b2sdk/bucket.py

```diff
@@ -292,6 +292,36 @@ def list_parts(self, file_id, start_part_number=None, batch_size=None):
         """
         return self.api.list_parts(file_id, start_part_number, batch_size)
 
+    def list_file_versions(self, file_name, fetch_count=None):
+        """
+        Lists all of the versions for a single file.
+
+        Depending on a Bucket's settings many versions may exist for the same filename.
+        This lists all of the versions present in the Bucket.
+
+        :param str file_name: the name of the file to list.
+        :param int,None fetch_count: how many entries to list per API call or ``None`` to use the default. Acceptable values: 1 - 1000
+        :rtype: generator[b2sdk.v1.FileVersionInfo]
+        :returns: generator of file_version_info
+        """
+        start_file_id = None
+        session = self.api.session
+        while True:
+            response = session.list_file_versions(
+                self.id_, file_name, start_file_id, fetch_count, file_name
+            )
+
+            for entry in response['files']:
+                file_version_info = FileVersionInfoFactory.from_api_response(entry)
+                if file_version_info.file_name != file_name:
+                    # All versions for the requested file name have been listed.
+                    return
+                yield file_version_info
+            if response['nextFileName'] != file_name:
+                # Unless there are more files with the same name we can stop here.
```

Comment:
a test to hit that condition is missing

Reply:
I've pushed the latest patch rebased to master but the test I wrote for this case hit some issues so I've commented it out. I would like some advice on how to proceed.

In order to test this condition the number of versions of a file must match the fetch count and the expectation would be that `nextFileName` would then be the name of the next file (or at least something that sorts later than the current file). However the RawSimulator doesn't behave this way, the `nextFileName` it returns is always the same name as the last file returned in the results:
https://github.com/Backblaze/b2-sdk-python/blob/6d35959ccfb06c3888702d91f3ce8d699f24f8e0/b2sdk/raw_simulator.py#L520

On the surface that is harmless enough, it just makes testing that case difficult. However the nextFileId that RawSimulator returns is also problematic. In the test I wrote I upload a single file which gets the id '9999' so with a fetch_count of 1 the first request returns the only version and then RawSimulator returns a nextFileName matching the file name and a nextFileId of '10000': https://github.com/Backblaze/b2-sdk-python/blob/6d35959ccfb06c3888702d91f3ce8d699f24f8e0/b2sdk/raw_simulator.py#L521

So the next request we send the same start_file_name and a start_file_id of '10000'. As RawSimulator loops over its known files it finds our file with id 9999 but the IDs are strings and it uses a string comparison to decide whether to include the file ID or not: https://github.com/Backblaze/b2-sdk-python/blob/6d35959ccfb06c3888702d91f3ce8d699f24f8e0/b2sdk/raw_simulator.py#L514

The string '10000' sorts lower than the string '9999' so the second request returns the same file id as the first and the same nextFileName and nextFileId. So the current implementation of the new API would loop forever in this case.

I can try and patch the RawSimulator to resolve these issues but I'm not sure if it would make more sense to do that as a separate PR or how you'd prefer to proceed.

Reply:
If RawSimulator has a bug that disturbs your test, then we should fix it. I think extending/fixing it in a PR that needs the fix is appropriate.

The question is, how does the cloud server behave in this exact case? From what the b2 server docs say and from what you say, RawSimulator implementation is flawed, but before modifying it, please test the same case with the cloud server - I want to avoid a situation where the inconsistency is caused by an error in documentation of the cloud server while actually the behavior of RawSimulator and cloud server are the same.

Reply:
Good catch. Looks like the b2 server does return the same nextFileName as the last returned file, just with a later (and in this case non-existent) nextFileId. I'll update to work with that case.